### ISSUE TITLE
Issues with High-Order GP model

### ISSUE THREAD

**Issue by saitcakmak** [pytorch/botorch#644]
# 🐛 Bug

I ran into a few issues when using the new Higher-Order GP model.

- If using `Standardize` as `outcome_transform` the model does not warn you to use `FlattenedStandardize`, and instead this raises an error due to shape mismatch in the `forward` pass. I think it would be helpful to either raise a warning and convert the `Standardize` to `FlattenedStandardize` while constructing `HigherOrderGP`, or to raise an error and to tell user to use the `FlattenedStandardize` instead.
- If using `HigherOrderGP` as the model in an `MCAcquisitionFunction` without specifying the `sampler`, this raises the error "SobolQMCSampler only supports dimensions `q * o <= 1111`". I think a good fix for this would be to implement a check similar to https://github.com/pytorch/botorch/blob/4ad285eb4566e3b278ae484dde2a60bad79d0344/botorch/optim/initializers.py#L97-L105 when initializing the sampler in the constructor of `MCAcquisitionFunction`.
- `HigherOrderGP` seems to use a lot of memory. See code below.

I'd be happy to put up a PR for first two issues. Not sure how to attack the last one.

## To reproduce

** Code snippet to reproduce **
```python
import torch
from torch import Tensor
from botorch import fit_gpytorch_model
from botorch.acquisition import MCAcquisitionObjective, qExpectedImprovement
from botorch.models.higher_order_gp import HigherOrderGP, FlattenedStandardize
from botorch.optim import optimize_acqf
from botorch.sampling import IIDNormalSampler
from gpytorch import ExactMarginalLogLikelihood

device = "cuda" if torch.cuda.is_available() else "cpu"

num_outputs = 50

train_X = torch.rand(10, 2, device=device)
train_Y = torch.rand(10, num_outputs, device=device)
model = HigherOrderGP(
    train_X,
    train_Y,
    outcome_transform=FlattenedStandardize(output_shape=torch.Size([num_outputs]))
)
mll = ExactMarginalLogLikelihood(model.likelihood, model)
_ = fit_gpytorch_model(mll)


class SumSquaresObjective(MCAcquisitionObjective):
    def forward(self, samples: Tensor) -> Tensor:
        return samples.pow(2).sum(dim=-1)


f_star = SumSquaresObjective()(train_Y).max()

acqf = qExpectedImprovement(
    model=model,
    best_f=f_star,
    objective=SumSquaresObjective(),
    # sampler=IIDNormalSampler(512),
)

sol, _ = optimize_acqf(
    acq_function=acqf,
    bounds=torch.tensor([[0.], [1.]]).repeat(1, 2).to(device=device),
    q=1,
    num_restarts=10,
    raw_samples=40,
)
```

** Stack trace/error message **
```
/home/saitcakmak/anaconda3/envs/other-BO/lib/python3.8/site-packages/torch/nn/modules/container.py:435: UserWarning: Setting attributes on ParameterList is not supported.
  warnings.warn("Setting attributes on ParameterList is not supported.")
Traceback (most recent call last):
  File "/home/saitcakmak/PycharmProjects/other-BO/debug_etc/hogp_memory_bug.py", line 39, in <module>
    sol, _ = optimize_acqf(
  File "/home/saitcakmak/anaconda3/envs/other-BO/lib/python3.8/site-packages/botorch/optim/optimize.py", line 151, in optimize_acqf
    batch_initial_conditions = ic_gen(
  File "/home/saitcakmak/anaconda3/envs/other-BO/lib/python3.8/site-packages/botorch/optim/initializers.py", line 112, in gen_batch_initial_conditions
    Y_rnd_curr = acq_function(
  File "/home/saitcakmak/anaconda3/envs/other-BO/lib/python3.8/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/saitcakmak/anaconda3/envs/other-BO/lib/python3.8/site-packages/botorch/utils/transforms.py", line 198, in decorated
    return method(cls, X, **kwargs)
  File "/home/saitcakmak/anaconda3/envs/other-BO/lib/python3.8/site-packages/botorch/utils/transforms.py", line 169, in decorated
    return method(cls, X, **kwargs)
  File "/home/saitcakmak/anaconda3/envs/other-BO/lib/python3.8/site-packages/botorch/acquisition/monte_carlo.py", line 159, in forward
    samples = self.sampler(posterior)
  File "/home/saitcakmak/anaconda3/envs/other-BO/lib/python3.8/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/saitcakmak/anaconda3/envs/other-BO/lib/python3.8/site-packages/botorch/sampling/samplers.py", line 93, in forward
    samples = posterior.rsample(
  File "/home/saitcakmak/anaconda3/envs/other-BO/lib/python3.8/site-packages/botorch/posteriors/transformed.py", line 93, in rsample
    samples = self._posterior.rsample(
  File "/home/saitcakmak/anaconda3/envs/other-BO/lib/python3.8/site-packages/botorch/models/higher_order_gp.py", line 338, in rsample
    kinv_rhs = self.train_train_covar.inv_matmul(train_rhs)
  File "/home/saitcakmak/anaconda3/envs/other-BO/lib/python3.8/site-packages/gpytorch/lazy/lazy_tensor.py", line 963, in inv_matmul
    return func.apply(self.representation_tree(), False, right_tensor, *self.representation())
  File "/home/saitcakmak/anaconda3/envs/other-BO/lib/python3.8/site-packages/gpytorch/functions/_inv_matmul.py", line 51, in forward
    solves = _solve(lazy_tsr, right_tensor)
  File "/home/saitcakmak/anaconda3/envs/other-BO/lib/python3.8/site-packages/gpytorch/functions/_inv_matmul.py", line 15, in _solve
    return lazy_tsr.cholesky()._cholesky_solve(rhs)
  File "/home/saitcakmak/anaconda3/envs/other-BO/lib/python3.8/site-packages/gpytorch/lazy/triangular_lazy_tensor.py", line 62, in _cholesky_solve
    res = self._tensor._cholesky_solve(rhs=rhs, upper=upper)
  File "/home/saitcakmak/anaconda3/envs/other-BO/lib/python3.8/site-packages/gpytorch/lazy/non_lazy_tensor.py", line 28, in _cholesky_solve
    return torch.cholesky_solve(rhs, self.evaluate(), upper=upper)
RuntimeError: CUDA out of memory. Tried to allocate 19.07 GiB (GPU 0; 5.80 GiB total capacity; 283.91 MiB already allocated; 2.26 GiB free; 462.00 MiB reserved in total by PyTorch)
```

## Expected Behavior

I'd expect it to use significantly less memory. If I replace `HigherOrderGP` with a `SingleTaskGP` (with 50 outputs), it uses less than 2GB memory.

## System information

**Please complete the following information:**
Botorch: Master 0.3.4.dev7+geb6cd5b8 
GPyTorch: 1.3.0
PyTorch: 1.7.1
OS: Ubuntu 20.04

**Comment by saitcakmak** [comment]
cc @wjmaddox @Balandat

**Comment by Balandat** [comment]
Re the standardization: It makes sense to warn the user. We can explicitly check for `FlattenedStandarize` for now, at a later point we may want to have some kind of base class that we can derive HOGP-compatible transforms from. 

Re the sampler: Implementing the fallback makes a lot of sense. Note that I have a PR up to increase the maximum dimension to 21201: https://github.com/pytorch/pytorch/pull/49710

Re the memory footprint: This should definitely not blow up like this. The lazy tensors that occur there should be `KroneckerProductLazyTensors` so that for solving the system we can use efficient structure exploiting linear algebra under the hood. Not quite sure what exactly causes this to not fire, @wjmaddox probably has a better idea. 
What version of gpytorch are you on? I believe you'll need this PR: which is not part of 1.3.0: https://github.com/cornellius-gp/gpytorch/pull/1394. We also might do some "smart" things for selecting cholesky solves based on the data covariance size in BoTorch that could cause the issues here.

**Comment by saitcakmak** [comment]
> What version of gpytorch are you on? I believe you'll need this PR: which is not part of 1.3.0: cornellius-gp/gpytorch#1394. 

I just created a clean environment from master, still trying to allocate 20+ GB with the above script (both on cpu and cuda).

**Comment by saitcakmak** [comment]
> Re the sampler: Implementing the fallback makes a lot of sense. Note that I have a PR up to increase the maximum dimension to 21201: pytorch/pytorch#49710

Looks like we need `model.posterior(...).event_shape[-2:]` for this. Is there an easy way of getting this without actually calling `model.posterior(X).event_shape[-2:]` with some dummy `X`? A particular choice would be `X = model.train_inputs[0][-1:]`.

Just for reference, this is what I'm thinking of replicating: https://github.com/pytorch/botorch/blob/4ad285eb4566e3b278ae484dde2a60bad79d0344/botorch/sampling/samplers.py#L291-L297

**Comment by Balandat** [comment]
I might have been a little premature in my response here - this model uses Matheron's rule to sample from the posterior. This involves sampling from the model prior, and then sampling independent noise to adjust those samples to generate samples from the posterior. We're writing something up on this right now, but it doesn't exactly fit the existing BoTorch framework. 

Regardless, it will depend on the shape of `X` what the overall dimension is. Specifically, the effective dimension for the sampler is `q*num_outputs`, where `q = X.shape[-2]` and  `num_outputs` is the total output dimension, i.e. the length of the vectorized output.

We could add a warning to `MCAcquisitionFunction.__init__`, but I'm not sure we need to do this right now - the HOGP model is still experimental and we probably need to iron out a few kinks until we suggest this for general use.

**Comment by wjmaddox** [comment]
I'll take a more detailed look later but i would think that something strange is going on in the construction of `train_rhs`, as `train_train_covar` should be a KroneckerProductLT of size 500 (50 x 10). 

I also didnt test out the HigherOrderGP in the case of a single output dimension, as it's really most useful for multiple output dimensions (50 x k) for example, but that would likely just increase the memory overhead.

**Comment by saitcakmak** [comment]
I think you're onto something there. If I use `10 x 50` or `10 x 1 x 50` `train_Y`, I get OOM error. But if I use `10 x 10 x 50` `train_Y`, it works. Maybe it needs an output shape with at least two elements `>1` for `KroneckerProductLT` to kick in?

**Comment by wjmaddox** [comment]
Looking into it a bit more deeply, the memory error is real in the sense that the matmul itself does generally fail
```
rhs = torch.randn(512, 40, 550, 1).cuda()
tensor = torch.randn(40, 550, 550).cuda()
tensor.matmul(rhs) # this also OOMs
```
The components are the 512 base samples, the 40 restarts and the 500 "training inputs" + 50 "test points". Setting `max_cholesky_size(0)` while optimizing the acquisition function resolves the issue -- presumably because this matmul doesn't happen.

But, I don't know why giving a SingleTaskGP 500 training inputs and trying to compute 50 posterior points with 512 * 40 posterior samples would work instead, which is what it ought to reduce to
```
train_X = torch.rand(500, 2, device=device)
train_Y = torch.rand(500, 1, device=device)
model = SingleTaskGP(
    train_X,
    train_Y,
    # outcome_transform=FlattenedStandardize(output_shape=torch.Size([num_outputs]))
)

test_x = torch.rand(550, 2, device=device)
base_samples = torch.randn(512, 40, 550, 1, device=device)

with gpytorch.settings.fast_computations(False):
    output = model.posterior(test_x).rsample(base_samples=base_samples, sample_shape=torch.Size((512, 40)))
# doesn't OOM
```

**Comment by saitcakmak** [comment]
So, sampling from `SingleTaskGP` in the last example uses `res = covar_root.matmul(base_samples) + self.loc.unsqueeze(-1)`, where `covar_root` is triangular lazy tensor with shape `550 x 550`. The extra dimension of size `40` is not there.

If I modify the model to be a batched GP with `40` batches, I get to the same line of code above with `covar_root: 40 x 550 x 550` and `base_samples: 40 x 550 x 512`. Though, this doesn't OOM either. Is it possible that `lazy.matmul` works more efficiently with the batches than the typical `matmul`?

**Comment by wjmaddox** [comment]
Thanks for tracking it down. This tells me that there's some reasonable memory improvements to be made in sampling the HOGP, which I'll look into.

**Comment by saitcakmak** [comment]
I simplified the script a little bit in case it helps (helped me :) )
```
    device = "cuda" if torch.cuda.is_available() else "cpu"

    num_outputs = [1, 50]

    train_X = torch.rand(10, 2, device=device)
    train_Y = torch.rand(10, *num_outputs, device=device)
    model = HigherOrderGP(
        train_X,
        train_Y,
    )

    X = torch.rand(40, 1, 2, device=device)
    posterior = model.posterior(X)
    base_samples = torch.rand(512, 40, 1050, device=device)
    samples = posterior.rsample(
        sample_shape=torch.Size([512]), base_samples=base_samples
    )
```

So, here with `num_outputs = [1, 50]` the `covar_root` in 
https://github.com/pytorch/botorch/blob/4ad285eb4566e3b278ae484dde2a60bad79d0344/botorch/models/higher_order_gp.py#L294-L296
is a `NonLazyTensor`. If I modify this with `num_outputs = [5, 50]`, `covar_root` becomes a `KroneckerProductLT`. In both cases, `covariance_matrix` is a `KroneckerProductLT`. This piece of code is the culprit:
https://github.com/cornellius-gp/gpytorch/blob/4dec144de87ae4b6900f4afffcd6f276455b6786/gpytorch/lazy/kronecker_product_lazy_tensor.py#L202-L204
It thinks the matrix is small, so it ignores the Kronecker structure.

**Comment by wjmaddox** [comment]
Thanks, that's what I'd assumed was going on when I plucked out the matmul failures, especially when setting `max_cholesky_size(0)` resolved it.

 The memory eating culprit that causes the OOMing is actually the trailing batch dimension in the sampling method as you pointed out --- broadly defined in the `_prepare_base_samples` method: https://github.com/pytorch/botorch/blob/4ad285eb4566e3b278ae484dde2a60bad79d0344/botorch/models/higher_order_gp.py#L201

Now, I'm just trying to ensure that getting the samples to be `batch x dim x num_samples` doesn't break the accuracy of the samples anywhere.

**Comment by saitcakmak** [comment]
Closed by #646 and #648
